Explicit casts to avoid conversion warnings [blocks: #2310]

tautschnig opened the pull request:
Visual Studio warns about signed/unsigned conversion and limited range of types.

smowton commented:
These really can't be silenced with an explicit typecast?

smowton commented:
Also a lot of these don't appear to be related to standard library calls?

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 5d40b601099f8b3ff9c5555a0723f06314e3d41b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78358680

tautschnig commented:
This is now re-implemented using explicit type casts.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: db43e7a4cab7f7dbd21ea6c38f74b72f40f9f166).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95906397

tautschnig committed:
Explicit casts to avoid conversion warnings: std::size_t

Visual Studio warns about signed/unsigned conversion and limited range of types.

tautschnig committed:
Explicitly cast bytes read from binaries to signed values

Visual Studio warns about signed/unsigned conversion and limited range of types.

tautschnig committed:
Explicitly cast returnvalue of tolower to char

Visual Studio warns about signed/unsigned conversion and limited range of types.

tautschnig committed:
std::distance returns a signed type, wrap in explicit cast

Visual Studio warns about signed/unsigned conversion and limited range of types.

tautschnig committed:
Explicit casts to avoid conversion warnings

Visual Studio warns about signed/unsigned conversion and limited range of types.

tautschnig committed:
Explicit casts to avoid conversion warnings: irept::set takes a signed long long

Visual Studio warns about signed/unsigned conversion and limited range of types.

tautschnig committed:
Use auto where type does not matter (and is likely signed)

This avoids conversion warnings with Visual Studio.

owen-mc-diffblue commented:
@tautschnig Should this be closed now that #3900 has been merged?

tautschnig commented:
@owen-jones-diffblue Many thanks for highlighting this - I need to check whether there are any changes left in this PR, certainly it does need a rebase.

tautschnig commented:
@owen-jones-diffblue Rebase done, there is still a fair number of changes left. Though arguably I should use `narrow_cast` instead of `static_cast`.

tautschnig commented:
Rebased and now uses `narrow_cast` throughout.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c332386494142389716c9ed24d42423cefb0dfc8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102222178

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: d96a3612557a55b681202bbce4267bfa00f67acc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104504752
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 13856db5c3f1e576bb526b8bce1085a010e07f0d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106408234

thk123 commented on src/ansi-c/literals/unescape_string.cpp:
```diff
@@ -44,15 +44,15 @@ std::basic_string<T> unescape_string_templ(const std::string &src)
   for(unsigned i=0; i<src.size(); i++)
   {
-    T ch=(unsigned char)src[i];
+    T ch = narrow_cast<T>(narrow_cast<unsigned char>(src[i]));
```
:question: Why narrow to unsigned char first? :mouse:

tautschnig replied:
This is to preserve the existing behaviour, as `T` might actually be wider than `unsigned char`. That said, I guess `narrow_cast` would fail in a case where that makes a difference, so I can probably get rid of it.

tautschnig replied:
Nested cast removed.

thk123 commented on src/ansi-c/literals/unescape_string.cpp:
```diff
@@ -44,15 +44,15 @@ std::basic_string<T> unescape_string_templ(const std::string &src)
       assert(i<src.size()); // backslash can't be last character
 
-      ch=(unsigned char)src[i];
+      ch = narrow_cast<T>(narrow_cast<unsigned char>(src[i]));
```
:mouse:

tautschnig replied:
Done.

thk123 commented on src/goto-programs/read_goto_binary.cpp:
```diff
@@ -99,7 +99,7 @@ static bool read_goto_binary(
         if(elf_reader.section_name(i)=="goto-cc")
         {
-          in.seekg(elf_reader.section_offset(i));
+          in.seekg(std::streamoff{elf_reader.section_offset(i)});
```
:question: Is this change meant to be here?

tautschnig replied:
Yes, because `section_offset` returns a `std::streampos` (which makes sense in other contexts).

thk123 reviewed (commented):
A couple of changes I think you should double check, but otherwise lgtm

tautschnig committed:
Remove nested cast

If the inner cast actually had any effect then that would amount to a
loss of precision, which in turn would necessarily fail with
narrrow_cast.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 36b9d0c42adfe9ae971ebb753825d55955ec26ce).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113768512

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: bf8ae774ff22f79fcf7420a1aa8ff359975ce9a9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113953221
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

TGWDB commented:
Closing due to age (no further comment on PR content), please reopen with rebase on develop if you intent to continue this work.
